add topology field and report NUMA node of device

cbf123 opened the pull request:
In order to allow the pod scheduling to be topology-aware, we need to
report the NUMA node that each device is associated with.

cbf123 commented:
This is the plugin side of the work being done in kubernetes to add a Topology Manager.  The corresponding API change in Kubernetes is tracked under https://github.com/kubernetes/kubernetes/pull/74423

cbf123 commented:
Note...I just copied the api.pb.go and api.proto files from the kubernetes master branch.  If this is not the proper way to do this please let me know.

ahalim-intel commented:
Hi @cbf123 thanks for the PR. We use go vendor with glide to manage dependencies for this project. Copy pasting dependencies inside the vendor don't work. In fact, the device plugin apis dependencies( v1beta1/api.pb.go) are already included in the project. All you need to do is update the glide with correct version/refs. Since the Topology manager is not fully merged yet, we do not have a commit id to update the vendor. This is why we decided to wait  until the upstream k8s is ready for this feature.

cbf123 commented:
@ahalim-intel Hi, thanks for the info.  The device plugin API change merged a couple weeks ago, doesn't that give a commit id?  Or do we need to wait for the complete feature?

ahalim-intel commented on pkg/resources/pciNetDevice.go:
```diff
@@ -56,9 +56,15 @@ func NewPciNetDevice(pciDevice *ghw.PCIDevice, rFactory types.ResourceFactory) (
 	mnt := infoProvider.GetMounts(pciAddr)
 	env := infoProvider.GetEnvVal(pciAddr)
 	rdmaSpec := rFactory.GetRdmaSpec(pciDevice.Address)
+        nodenum = utils.GetDevNode(pciAddr)
```
compilation error.
it should be:
```
nodenum := utils.GetDevNode(pciAddr)
```

cbf123 replied:
thanks.  updated.

ahalim-intel commented on pkg/utils/utils.go:
```diff
@@ -115,6 +115,21 @@ func GetVFconfigured(pf string) int {
+	if err != nil {
+		return -1
+	}
+ 	node := bytes.TrimSpace(node)
```
Compilation error: variable 'node' already declared at L:121

ahalim-intel commented:
> @ahalim-intel Hi, thanks for the info. The device plugin API change merged a couple weeks ago, doesn't that give a commit id? Or do we need to wait for the complete feature?

Can you find that commit id and I can give it a try locally.

cbf123 commented:
Thanks for the reviews, I think I've addressed all the issues.

Also, I checked the pkg/kubelet/apis/deviceplugin/v1beta1 and pkg/kubelet/apis/pluginregistration/v1 subpackages of kubernetes and the only change in the new git commit ID is the addition of the new topology field.

cbf123 commented:
Added some tests.

cbf123 commented:
It looks like I need to do a "glide update" or equivalent for the APIs pulled in from Kubernetes.  What's the proper way to do this?  If I run "glide update" it ends up changing a whole bunch of unrelated stuff which causes "make test" to break in areas I didn't touch.

Also, I guess I shouldn't be specifying the commit ID in the glide.lock file manually, but rather should be putting it in the glide.yaml file?   How does that work when "k8s.io/kubernetes" isn't a git repo?

ahalim-intel commented:
> It looks like I need to do a "glide update" or equivalent for the APIs pulled in from Kubernetes. What's the proper way to do this? If I run "glide update" it ends up changing a whole bunch of unrelated stuff which causes "make test" to break in areas I didn't touch.
> 
> Also, I guess I shouldn't be specifying the commit ID in the glide.lock file manually, but rather should be putting it in the glide.yaml file? How does that work when "k8s.io/kubernetes" isn't a git repo?

After adding the current version you need to run `glide update -v`.  This may result in updating other dependencies as well and not guaranteed to work. Sometime you may need to update other dependencies version too. It's tricky to get a common versions of package that one or more packages depend on.

One other concern I have, once we update the K8s api version will device plugin be compatible with older version of K8s? Have you tested the device plugin with this changes with an older version of kubelet?

igsilya commented on pkg/utils/utils_test.go:
```diff
@@ -172,6 +172,28 @@ var _ = Describe("In the utils package", func() {
+			Expect(GetDevNode(pciAddr)).To(Equal(expected))
+		},
+		Entry("reading the device path fails", &FakeFilesystem{}, "0000:00:00.0", -1),
+		Entry("converting the VFs number to integer fails",
```
Leftovers from copy-pasting.

cbf123 replied:
updated in most recent version.

igsilya commented on pkg/utils/utils_test.go:
```diff
@@ -172,6 +172,28 @@ var _ = Describe("In the utils package", func() {
+			},
+			"0000:00:00.1", 1,
+		),
+	)
```
It'll be good to have a test for reading '-1' from 'numa_node' as it's the common  and very frequent case.

cbf123 replied:
updated in most recent version

ahalim-intel commented on pkg/utils/utils.go:
```diff
@@ -185,6 +185,21 @@ func GetSriovVFcapacity(pf string) int {
 	return numvfs
 }
 
+// GetDevNode returns the numa node of a PCI device, -1 if none is specified or error.
```
> Returning '-1' on error doesn't look like a good idea. This is a normal case when kernel was built without NUMA support and in this case where will be no 'numa_node' knob or it will always return '-1' as a result. I think that it's better to return 0 in these cases. Otherwise you'll not be able to ever start a pod with 'strict' policy on a node without NUMA support (like inside the VM).

I agree, we need to cognizant of non-numa systems. Let's verify how the Topology manager handles the non-numa scenario and what do they expect from device plugin in response from such system.

igsilya replied:
kubernetes uses cadvisor to get topology information. cadvisor parses /proc/cpuinfo for 'physical id' and counts number of nodes this way. So, for non-NUMA systems there be single node with id 0. For arches without 'physical id' in cpuinfo it parses /sys/bus/cpu/devices for 'nodeX' files. On failure it reports node id 0.
ref: https://github.com/google/cadvisor/blob/d200c9f558b0a8051d728fd8bb74d0469c6a36e1/machine/machine.go#L194

cbf123 replied:
I'll go back and ping the folks working on the k8s side of things as to what the expectations are for the case where the kernel returns "-1" (due to the device not having numa affinity) or the host kernel was built without numa awareness.

I actually had intentionally used -1 on error to match what the kernel returns, but maybe they want nil instead.  Returning 0 is definitely not the right thing to do, as in a VM it's possible to have multiple virtual NUMA nodes but devices with numa_node=-1.

cbf123 replied:
Okay, based on the reply at https://github.com/kubernetes/kubernetes/pull/80570#issuecomment-523125339 the expectation is that if we cannot determine a NUMA node then we should simply not set the field.  I'll update the code to reflect that.

Given this, I think it actually makes sense to preserve the "-1" in GetDevNode() and then check for that to decide whether or not to set the "topology" field at all.

klueska replied:
Please let me know if you see any issues with this approach.

adrianchiris replied:
setting the field to -1 or not setting it is just semantics as PB definition has this field defined with a default of -1.

it would be useful to state what -1 means to devicemanager in PB API definition & relevant documentation.

klueska replied:
To expand on what happens in the `devicemanager`/`TopologyManager` in this case:

From the `devicemanager`'s perspective, having even one device with `Topology.Node` unset, will result in it returning a `nil` for the `[]TopologyHint` for that device type.

When the `TopologyManager` sees a `nil` `[]TopologyHint` for a given resource type, it treats that as a "don't care" alignment for that resource type and generates a single hint with a full socket mask for that resource type (indicating allocations can occur from any socket for that resource type). This will then be combined with hints from other resource types to generate a "merged" hint with a common alignment amongst all resource types.

When the `strict` policy (recently renamed as `restricted`) performs its merge strategy, it will only admit the pod if resource alignment on a _single_ NUMA node can be satisfied (i.e. if even one resource from any resource type needs to be pulled from a second NUMA node, the pod will not be admitted).

So in the case being discussed here (i.e. of a VM only seeing a single NUMA node and returning -1 as its Node ID), we should still be OK (even with the `restricted` policy).

The `devicemanager` will return `nil` for its hints on all resource types, the `restricted` policy will turn this into a hint with a filled socketmask with only one bit set (because there is only one socket on the machine) and it will admit the pod because alignment can (and by construction must be) satisfied on that single socket in the mask.

ahalim-intel replied:
@klueska Thanks for explaining the behavior. I think from device plugin point of view it only make sense to add `TopologyInfo` if we able get valid NUMA info from the platform. Otherwise we should leave this field to be `nil` and let device manager decide how this information is communicated toTopologyManager.

klueska replied:
Agreed

adrianchiris replied:
> From the `devicemanager`'s perspective, having even one device with `Topology.Node` unset, will result in it returning a `nil` for the `[]TopologyHint` for that device type.
> 
> When the `TopologyManager` sees a `nil` `[]TopologyHint` for a given resource type, it treats that as a "don't care" alignment for that resource type and generates a single hint with a full socket mask for that resource type (indicating allocations can occur from any socket for that resource type). This will then be combined with hints from other resource types to generate a "merged" hint with a common alignment amongst all resource types.

So, in case we are in a NUMA system, Numa affinity is required and we failed to get the NodeID for whatever reason for a **single** device in a (SR-IOV) resource then it may lead to a TopologyHint that may not be satisfiable by devicemanager for that specific resource ?

klueska replied:
As it's implemented right now, if any single device for a given device type does not have the the `Topology` field set (even if all of the other devices of the same type do have it set), then the `devicemanager` will return `nil` for its list of hints for that device type (resulting in a "don't care" alignment for that device type).

klueska replied:
An alternative would be to set the socketmask to match the devices that do have the `Topology` field set and just ignore the devices that don't have it set.

In the case of the `restricted` policy, this would result in only admitting pods that matched alignment against the devices it could read the topology from (i.e. the device without `Topology` would never be available for allocation). In the case of the `best-effort` policy, the device without `Topology` would end up being a sort of "last resort" device assuming no other alignment could be satisfied.

Are these semantics preferred to the existing ones?

adrianchiris replied:
That's what i had in mind.

klueska replied:
Upon reflection, I think these "newer" semantics are more correct than the old ones (even if they aren't perfect).

In the existing implementation, both `restricted´ and `best-effort` policies have the ability to admit pods that may not have true alignment (because falling back to "don't care" when no `Topology` field is set, essentially makes devices of that type appear to be aligned on every socket and thus allocatable). This goes against the entire purpose of the `restricted` policy and would likely lead to much confusion down the road.

I am going to update https://github.com/kubernetes/kubernetes/pull/80570 to reflect the updated semantics.

cbf123 replied:
If a host kernel is built without CONFIG_NUMA I think the `numa_node` file for all devices on that host will simply be missing and so there would be no `Topology` field.

If an pod really cares about NUMA-affinity, it would kind of defeat the purpose to put it on a host that is not NUMA-aware.

As such, I think it would make sense that devices without `Topology` wouldn't match for the `strict` policy.

cbf123 commented:
> One other concern I have, once we update the K8s api version will device plugin be compatible with older version of K8s? Have you tested the device plugin with this changes with an older version of kubelet?

I have not tested it with an older version of kubelet.  It should just ignore the unknown field, right?

cbf123 commented:
In the most recent iteration I updated glide.yaml to the kubernetes commit that updated the device plugin API, ran "glide update -v" and then backed out all the vendor changes except the kubernetes ones.  Hopefully this is acceptable.

zshi-redhat commented:
we should verify and confirm that this works with current kubernete version before merging.

> > One other concern I have, once we update the K8s api version will device plugin be compatible with older version of K8s? Have you tested the device plugin with this changes with an older version of kubelet?
> 
> I have not tested it with an older version of kubelet. It should just ignore the unknown field, right?

I tried with v1.15.2, it works fine. 
Version details:
`Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.2", GitCommit:"f6278300bebbb750328ac16ee6dd3aa7d3549568", GitTreeState:"clean", BuildDate:"2019-08-05T09:23:26Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.2", GitCommit:"f6278300bebbb750328ac16ee6dd3aa7d3549568", GitTreeState:"clean", BuildDate:"2019-08-05T09:15:22Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}`

zshi-redhat commented:
> In the most recent iteration I updated glide.yaml to the kubernetes commit that updated the device plugin API, ran "glide update -v" and then backed out all the vendor changes except the kubernetes ones. Hopefully this is acceptable.

I also did a local `glide update -v` with this PR applied, it results in a different glide.lock with other packages(exclude kubernetes pkg) being pinned to newer commit versions.
I think the vendor pkg should be ok for this PR as long as binary build passes. we could do another glide update to bring latest code from other depend pkgs.
One minor comment is if there will be additional update on this PR, would you please split glide update and real code change in separate commits?

cbf123 commented:
I can happily split them into two commits...I only did them as one because the last glide update was also a single commit. :)

zshi-redhat commented:
Although it's not supported using device plugin in VM yet, but I'd like to understand how would this be useful in a VM environment, will the numa info retrieved within VM reflects the fact of VF locality? or is it not useful at all because NUMA is configured by VM provider(e.g., OpenStack) in VM spec.

zshi-redhat commented:
> In general it's not useful to report the numa_node in a VM, because there is no guarantee that the NUMA node as seen by the guest has any relationship to the underlying physical topology on the host.
> 
> As a thought experiment, consider the case where a single-NUMA-node VM is running on host NUMA node 1, and is associated with a VF from a PF located on host NUMA node 2. Within the guest there is only one NUMA node so the only thing it could report (other than -1) for the VF numa_node would be NUMA node 0...but that's clearly wrong from a host topology perspective.

Right, in the VM case. we probably don't want to use NUMA feature as:
1) it might report a wrong hints
2) NUMA affinity may already be guaranteed by infrastructure, e.g. OpenStack where VMs are created on.

It may not be a good idea to have devcie plugin detect whether it's running in a VM environment or not, would it make sense to have a switch to turn off reporting of NUMA info?

zshi-redhat commented:
Would you please also populate `numa` field of `pciNetDevice` structure in `pkg/resources/pciNetDevice.go` file. It might be useful when other parts of plugin code to reference the numa info.

cbf123 commented:
So in my experience, when running in a VM (even using KVM) the virtual devices report a numa_node of -1.  Does anyone know what gets reported for pci-passthrough devices?

I've updated the code to set the existing "numa" field.

cbf123 commented:
Based on [this](https://www.berrange.com/posts/2017/02/16/setting-up-a-nested-kvm-guest-for-developing-testing-pci-device-assignment-with-numa/) it looks like if you jump through a bunch of hoops it's possible to get a VM to report numa-affinity for PCI devices.   But I think that if you jump through all those hoops it's probably because you want it to behave like a real system and therefore would want to use the reported numa_node "normally".

As such, I'm not sure we need a switch to control reporting  NUMA or not.

igsilya commented:
> Based on [this](https://www.berrange.com/posts/2017/02/16/setting-up-a-nested-kvm-guest-for-developing-testing-pci-device-assignment-with-numa/) it looks like if you jump through a bunch of hoops it's possible to get a VM to report numa-affinity for PCI devices. But I think that if you jump through all those hoops it's probably because you want it to behave like a real system and therefore would want to use the reported numa_node "normally".

I agree with that. We tested the pci numa affinity for VMs manually using steps similar to described in above article. Most of VM providers doesn't care about pci numa affinity, but it's possible to configure by hands and if you're configuring it by hands, you're most probably know what you're doing. Even without real passthrough devices, this functionality is highly useful for kubernetes testing.

> As such, I'm not sure we need a switch to control reporting NUMA or not.

Anyway, '-1' will be converted to missing Topology field in a device structure, so topology manager will ignore this devices while making decisions.

igsilya commented on pkg/resources/pciNetDevice.go:
```diff
@@ -82,7 +91,7 @@ func NewPciNetDevice(pciDevice *ghw.PCIDevice, rFactory types.ResourceFactory) (
 		mounts:      mnt,
 		env:         env,
-		numa:        "", // TO-DO: Get this using utils pkg
+		numa:        strconv.Itoa(nodenum),
```
I know, that this field is never actually used, but shouldn't this be `""` for a `-1` case?

cbf123 replied:
Sure, I can do that.  There's some merit to keeping the current behaviour in the no-NUMA case.

zshi-redhat commented:
> Based on [this](https://www.berrange.com/posts/2017/02/16/setting-up-a-nested-kvm-guest-for-developing-testing-pci-device-assignment-with-numa/) it looks like if you jump through a bunch of hoops it's possible to get a VM to report numa-affinity for PCI devices. But I think that if you jump through all those hoops it's probably because you want it to behave like a real system and therefore would want to use the reported numa_node "normally".

Assume a multi-numa VM node, if user doesn't care about numa affinity with SR-IOV device inside VM, then would it be a matter to still report numa_node info of that device?  for example, depends on topology manager policy used (best effort, restrict etc), will kubelet only schedule pods to cpus on the selected numa node(based on reported numa info from device plugin), leaving other cpus on different numa node unused ?

> 
> As such, I'm not sure we need a switch to control reporting NUMA or not.

adrianchiris commented:
> Assume a multi-numa VM node, if user doesn't care about numa affinity with SR-IOV device inside VM, then would it be a matter to still report numa_node info of that device? 

To my understanding, 
If a resource doesnt care about NUMA it must not report it.
Reporting NUMA will affect Topology Hint generation by topology manager (which is a policy specific behaviour) and device allocation by device manager.

zshi-redhat commented:
> > Assume a multi-numa VM node, if user doesn't care about numa affinity with SR-IOV device inside VM, then would it be a matter to still report numa_node info of that device?
> 
> @zshi-redhat
> To my understanding,
> If a resource doesnt care about NUMA it must not report it.

Then question is how would we allow user not reporting the numa hints. should device plugin be able to disable this report or we can just rely on kubelet config to not enable numa.

> Reporting NUMA will affect Topology Hint generation by topology manager (which is a policy specific behaviour) and device allocation by device manager.

cbf123 commented:
In the current implementation there is a config option for kubelet to select between "best-effort" or "strict" affinity.  With "strict" affinity it will only allow a container to run if all resources are on the same NUMA node.  With "best-effort" it will try to put them on the same NUMA node but will still allow it to start up even if resources are spread across NUMA nodes.

Personally I think it'd make more sense for the strict/best-effort flag to be set on the container rather than on kubelet.

klueska commented:
> Personally I think it'd make more sense for the strict/best-effort flag to be set on the container rather than on kubelet.

This was discussed here:
https://docs.google.com/document/d/1g5Aqa0BncQGRedSJH0TJQWq3mw3VxpJ_ufO1qokJ1LE/edit?disco=AAAADW5cXY8

> In any case, I don't think we need a flag to control exporting NUMA information from the plugin. If the admin doesn't care about NUMA, they can set "best-effort" in kubelet.

Agreed. Or just not enable the TopologyManager at all.

zshi-redhat commented:
> In the current implementation there is a config option for kubelet to select between "best-effort" or "strict" affinity. With "strict" affinity it will only allow a container to run if all resources are on the same NUMA node. With "best-effort" it will try to put them on the same NUMA node but will still allow it to start up even if resources are spread across NUMA nodes.
> 
> Personally I think it'd make more sense for the strict/best-effort flag to be set on the container rather than on kubelet.
> 
> In any case, I don't think we need a flag to control exporting NUMA information from the plugin. If the admin doesn't care about NUMA, they can set "best-effort" in kubelet.

Makes sense, `best-effort` policy can be used if user doesn't care NUMA.

cbf123 commented:
@zshi-redhat it looks like you switched to using go.mod but you kept the existing glide.lock  and glide.yaml files even though they no longer represent what's in the vendor directory.  That seems not quite right.

In any case, this patch can't be rebased due to a change in the vendor directory.  I need to rebuild it.  I assume I need to use go.mod to specify the new k8s version now?

zshi-redhat commented:
> @zshi-redhat it looks like you switched to using go.mod but you kept the existing glide.lock and glide.yaml files even though they no longer represent what's in the vendor directory. That seems not quite right.

Yes, I switched to go.mod as the glide update is not working well as seen in the PR and the update also brings latest grpc version which fixes several CVE bugs(using `glide update -v` doesn't resolve the dependency correctly).
I will remove glide config files in a separate patch.

> 
> In any case, this patch can't be rebased due to a change in the vendor directory. I need to rebuild it. I assume I need to use go.mod to specify the new k8s version now?

use go 1.12 version and run :
1)  apply `add topology field and report NUMA node of device` commit on top of current master
2) `export GO111MODULE=on` if not set.
4)  `make build`  to update the go.sum
5) `go mod tidy` : this removes unused checksum from go.sum
6) `go mod vendor` : this update the local vendor pkgs

zshi-redhat commented:
> @zshi-redhat it looks like you switched to using go.mod but you kept the existing glide.lock and glide.yaml files even though they no longer represent what's in the vendor directory. That seems not quite right.
> 
> In any case, this patch can't be rebased due to a change in the vendor directory. I need to rebuild it. I assume I need to use go.mod to specify the new k8s version now?

@cbf123 I'm removing glide files in this PR: https://github.com/intel/sriov-network-device-plugin/pull/172

ahalim-intel commented:
> @cbf123 I'm removing glide files in this PR: #172

Update: 
We've decided to split out ^above PR and create separate PR to address the glide and travis-ci related changes. The glide files are removed in this PR #173

cbf123 commented:
> 
> use go 1.12 version and run :
> 
>     1. apply `add topology field and report NUMA node of device` commit on top of current master
> 
>     2. `export GO111MODULE=on` if not set.
> 
>     3. `make build`  to update the go.sum

This results in an error using go 1.12.9:

../../pkg/resources/pciNetDevice.go:66:12: apiDevice.Topology undefined (type *v1beta1.Device has no field or method Topology)
../../pkg/resources/pciNetDevice.go:66:25: undefined: v1beta1.TopologyInfo
make: *** [/home/cfriesen/devel/sriov-network-device-plugin/build/sriovdp] Error 2


I also get an error if I just try running "go get k8s.io/kubernetes":
...initial bit snipped...
go: k8s.io/cluster-bootstrap@v0.0.0: unknown revision v0.0.0
go: k8s.io/kube-proxy@v0.0.0: unknown revision v0.0.0
go: k8s.io/code-generator@v0.0.0: unknown revision v0.0.0
go: k8s.io/metrics@v0.0.0: unknown revision v0.0.0
go: k8s.io/kubelet@v0.0.0: unknown revision v0.0.0
go: k8s.io/legacy-cloud-providers@v0.0.0: unknown revision v0.0.0
go: error loading module requirements

cbf123 commented:
According to https://github.com/golang/go/issues/32776 it might be a problem with specifying k8s.io/kubernetes as a dependency:

> k8s.io/kubernetes is not primarily intended to be consumed as a module. Only the published subcomponents are (and go get works properly with those).

zshi-redhat commented:
> > use go 1.12 version and run :
> > ```
> > 1. apply `add topology field and report NUMA node of device` commit on top of current master
> > 
> > 2. `export GO111MODULE=on` if not set.
> > 
> > 3. `make build`  to update the go.sum
> > ```
> 
> This results in an error using go 1.12.9:
> 
> ../../pkg/resources/pciNetDevice.go:66:12: apiDevice.Topology undefined (type *v1beta1.Device has no field or method Topology)
> ../../pkg/resources/pciNetDevice.go:66:25: undefined: v1beta1.TopologyInfo
> make: *** [/home/cfriesen/devel/sriov-network-device-plugin/build/sriovdp] Error 2
> 
> I also get an error if I just try running "go get k8s.io/kubernetes":
> ...initial bit snipped...
> go: k8s.io/cluster-bootstrap@v0.0.0: unknown revision v0.0.0
> go: k8s.io/kube-proxy@v0.0.0: unknown revision v0.0.0
> go: k8s.io/code-generator@v0.0.0: unknown revision v0.0.0
> go: k8s.io/metrics@v0.0.0: unknown revision v0.0.0
> go: k8s.io/kubelet@v0.0.0: unknown revision v0.0.0
> go: k8s.io/legacy-cloud-providers@v0.0.0: unknown revision v0.0.0
> go: error loading module requirements

I hit the same issue.
I also tried to pin the latest commit of k8s.io/kubernetes and use go version 1.13, neither works.
I don't know if we need to wait until 1.16 gets released which may resolve the conflict?
Or as a temporary solution, copy the kubernetes vendor pkg manually like we did with glide and have all the build using local vendor pkgs instead of running go mod everytime.

cbf123 commented:
I've opened issue https://github.com/kubernetes/kubernetes/issues/82437 to track this, because it seems like a deficiency in the core kubernetes code.

zshi-redhat commented:
> I've opened issue [kubernetes/kubernetes#82437](https://github.com/kubernetes/kubernetes/issues/82437) to track this, because it seems like a deficiency in the kubernetes code.

@cbf123 hi, I chatted with @ahalim-intel yesterday and we agreed to revert the `go mod` support in favor of having this NUMA patch merged. Here is the revert change [PR](https://github.com/intel/sriov-network-device-plugin/pull/176).

With above change, we will continue to use glide for vendor pkg management and hold the two PRs that try to remove glide support(https://github.com/intel/sriov-network-device-plugin/pull/173 and https://github.com/intel/sriov-network-device-plugin/pull/172) until there is a solution to add kubernetes pkg as dependency with `go module`.

cbf123 commented:
So only the last two commits are really relevent...the third-last one was fixing up a bug in https://github.com/intel/sriov-network-device-plugin/pull/176

cbf123 commented:
I've updated the code...this time I had to update all the vendor packages.  If I only included the Kubernetes changes then the self-tests ran into grpc incompatibilities.

ahalim-intel commented:
Are there any more comments on this PR?

zshi-redhat commented:
Tested with `restricted` topology manager policy + `static` cpu manager policy, sr-iov pod admit works as expected.

kubernetes version:
``Client Version: version.Info{Major:"1", Minor:"17+", GitVersion:"v1.17.0-alpha.0.1511+124ed02c93b868", GitCommit:"124ed02c93b868837fe9cf8f357db7d234605a73", GitTreeState:"clean", BuildDate:"2019-09-18T01:20:15Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17+", GitVersion:"v1.17.0-alpha.0.1511+124ed02c93b868", GitCommit:"124ed02c93b868837fe9cf8f357db7d234605a73", GitTreeState:"clean", BuildDate:"2019-09-18T01:18:03Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}``

zshi-redhat commented:
/lgtm

cbf123 commented:
Thanks everyone.
